#1730 cargo console

vlandemart opened the pull request:
### Purpose
Cargo console. Fixes #1730

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### Notes:
For now, only 3 tabs are implemented: Supplies (with all possible things you can order), cart and status. Nice stuff like bounties screen and separate request console will be done sometime.
Also, I couldn't come up with nice bounty system, so for now it's just a placeholder giving 50 credits for every ObjectBehaviour on shuttle.
Supplies data is held in "CargoData" scriptable object and loaded in CargoManager on GUI_Cargo.InitServer()
!!!CONTAINS SCENE CHANGE!!!

vlandemart committed:
Initial commit

vlandemart committed:
Various after-merge fixes

Changed NetTabs for NetPages, fixed several errors after merging with latest dev

vlandemart committed:
Rewrote stuff to NetPages

Tabs are now netsynced (Supply tab isnt working)

vlandemart committed:
Fixed build NRE

*Changed Managers.prefab so console doesn't give NRE on build
*Supplies tab is working now

vlandemart committed:
Categories

Implemented Categories in Supplies tab

vlandemart committed:
Status tab

*Nicer status tab with more information, message window and timer
*Timer behaviour for cargo shuttle
*Separate file containing data (CargoData scriptavble object)

vlandemart committed:
Merge with develop

(removed my scene changes)

vlandemart committed:
Final commit (CONTAINS SCENE CHANGE)

*Mapped cargo console onto scene
*Implemented current shuttle status icon

vlandemart committed:
Fixed reversed scroll in messages

Changed scrollbar direction to "bottom to top" on CentcomMessages

vlandemart committed:
Removed unnecessary Debug.Logs

fomalsd commented on UnityProject/Assets/Scripts/Consoles/CargoConsole.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class CargoConsole : NetworkTabTrigger
```
hm, it's kinda identical to `NetworkTabTrigger`. might be a good idea to make `NetworkTabTrigger` non-abstract and remove this one (looks like design oversight)

fomalsd commented on UnityProject/Assets/Scripts/Managers/CargoManager.cs:
```diff
@@ -0,0 +1,255 @@
+		if (CentcomMessage == "")
+			CentcomMessage = "Bounty items recieved.\n";
+		//1 - quantity of items
+		CentcomMessage += "+" + credits + " credits: recieved " + "1" + " " + item.name + ".\n";
```
Some grouping would be nice in the future

fomalsd replied:
Also use string interpolation for things like this so they're more readable:
![GIF](https://user-images.githubusercontent.com/10403536/58466729-ca594100-8129-11e9-87c4-d93d4b8a01f1.gif)
And there's a typo

fomalsd replied:
Aand there's (Clone) text everywhere, maybe use `item.gameObject.ExpensiveName()` here instead

fomalsd commented on UnityProject/Assets/Scripts/UI/Cargo/GUI_CargoTab.cs:
```diff
@@ -0,0 +1,26 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class GUI_CargoTab : NetPage
```
would be better if it was called GUI_CargoPage, because it's confusing (as Tab is assumed to be the entire machine UI)

fomalsd reviewed (approved):
Very, very quality stuff!
There are several minor things I've noticed, but they can be fixed later as I'm inclined to merge this sooner because of scene and prefab changes

vlandemart commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -0,0 +1,155 @@
+		int height = 4;
+
+		Vector3Int spawnPos;
+		while (true)
```
Results in an endless loop which freezes the game if you orded too much crates. Will fix it in few hours.

vlandemart reviewed (commented):
There is a bug that will result in frozen game if you order more crates that shuttle can fit

vlandemart committed:
Various fixes

* Fixed endless while on Order spawn that resulted in game freeze
* Renamed sub-tabs into pages
* Now order spawns on leaving Centcom rather then arriving

fomalsd commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -138,18 +141,21 @@ private Vector3 GetRandomFreePos()
+			i++;
 		}
-		return spawnPos;
+		return Vector3.zero;
```
use `TransformState.InvalidPos` as null object, because `Vector3.zero` can be valid spawning position

vlandemart replied:
You mean `TransformState.HiddenPos`?

fomalsd replied:
yeah, HiddenPos

fomalsd commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -138,18 +141,21 @@ private Vector3 GetRandomFreePos()
 		Vector3Int spawnPos;
-		while (true)
+		//temporary max crates in one
+		while (i < 40)
```
why is there a hardcoded attempt limit? I mean, it can randomly choose to check same occupied position 40 times in a row and then give up

vlandemart replied:
I just couldn't come up with nice way of getting random position without repeats.
I mean, i can make double array check, but it wouldn't spawn crates randomly scattered around.

fomalsd replied:
Before starting to spawn orders you could calculate a List of non-occupied positions (`shuttleMatrix.IsPassableAt(LocalPos)`/`MatrixManager.IsPassableAt(WorldPos)`) within shuttle bounds, shuffle it, put it into a Queue and then Dequeue valid positions from that queue

fomalsd commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -118,16 +118,19 @@ void UnloadCargo()
+	public bool SpawnOrder(CargoOrder order)
 	{
 		Vector3 pos = GetRandomFreePos();
+		if (pos == Vector3.zero)
```
check for `TransformState.InvalidPos` here, too

vlandemart committed:
Proper GetRandomFreePos() fix 

Available positions are now calculated before spawning objects and stored in list.

fomalsd commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -133,29 +145,44 @@ public bool SpawnOrder(CargoOrder order)
-			i++;
+			spawnPos = availableSpawnSlots[Random.Range(0, availableSpawnSlots.Count)];
+			availableSpawnSlots.Remove(spawnPos);
+			Debug.Log(availableSpawnSlots.Count + " empty");
```
```suggestion
			Debug.Log(availableSpawnSlots.Count + " empty", Category.NetUI);
```

vlandemart replied:
Forgot to delete testing Debug.Logs

fomalsd commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -12,6 +12,9 @@ public class CargoShuttle : MonoBehaviour
 	private int dockOffset = 23;
 	private Vector3 destination;
+	private List<Vector3Int> availableSpawnSlots = new List<Vector3Int>();
+	private int shuttleWidth = 2;
```
should have probably commented that it's `(cargoZoneWidth - 1) / 2` as it took me some time to figure out their meaning

fomalsd commented on UnityProject/Assets/Scripts/Shuttles/CargoShuttle.cs:
```diff
@@ -133,29 +145,44 @@ public bool SpawnOrder(CargoOrder order)
+		{
+			for (int j = -shuttleWidth; j <= shuttleWidth; j++)
+			{
+				pos = Vector3Int.RoundToInt(transform.position);
```
ideally it would be `pos = matrixMove.State.Position.RoundToInt()`, as `transform.position` isn't guaranteed to be accurate and only represents clientside lerping position (that can theoretically be way off)
You'll need to cache a reference to MatrixMove somewhere like OnEnable() though (or do lazy init with backing field), and it would be even safer if you placed `[RequireComponent(typeof(MatrixMove))]` attribute above CargoShuttle class definition

fomalsd reviewed (commented):
just a couple more tiny things before merge

vlandemart committed:
CargoShuttle fixes

* Added few comments
* Removed Debug.Logs
* Replaced transform.position with matrix position
